_posts/2021-08-12-Alias-Declarations-over-Typedefs-CPP.md

```diff
@@ -0,0 +1,118 @@
+---
+layout: post
+title: "Prefer Alias Declarations to Typedefs (Notes)"
+date: 2021-08-12
+desc: "Notes on Item 9, Chapter 3 of Effective Modern C++"
+keywords: "Blog, C++, Vectors, Notes"
+categories: [Blog, C++]
+tags: [Blog, C++]
+blog: [C++]
+published: true
+excerpt_separator: <!--more-->
+---
+
+---
+**NOTE**
+
+My notes on Chapter 3, Item 9 of Effective Modern C++ written by Scott Meyers.
+
+Some (or even all) of the text can be similar to what you see in the book, as these are notes: I've tried not to be unnecessarily creative with my words. :)
+
+---
+
+<!--more-->
+
+One solution to avoiding using long type names:
+
+```cpp
+// So C++98 like
+typedef std::unique_ptr<std::unordered_map<std::string, std::string>> UPtrMapSS;
+```
+
+C++11 also offers alias declarations:
+
+```cpp
+using UPtrMapSS = typedef std::unique_ptr<std::unordered_map<std::string, std::string>> 
+```
+
+Advantages of alias declarations over typedefs:
+
+1. For types involving function pointers, aliases are easier to read (for some people):
+    ```cpp
+    // typedef
+    typedef void (*FP)(int, const std::string&);
+
+    // alias declaration
+    using FP = void (*)(int, const std::string&);
+    ```
+2. Alias declarations can be templatized, but typedefs can not.
+    ```cpp
+    // MyAlloc is a custom allocator
+    template <typename T>
+    using MyAllocList = std::list<T, MyAlloc<T>>;
+
+    MyAllocList<Widget> lw; // will create std::list<Widget, MyAlloc<Widget>>
+    ```
+
+    vs: typedef, a hack way:
+
+    ```cpp
+    // templatized struct here
+    template <typename T>
+    struct MyAllocList {
+        typedef std::list<T, MyAlloc<T>> type;
+    };
+
+    MyAllocList<Widget>::type lw;
+    ```
+3. In case you want to use a type specified by template parameter, with typedefs it gets complex:
```

Comment:
Just mention, using the `MyAllocList` above. (For one second, I was trying to see) (Maybe I am mindless reader sometimes, but pointing that the definition is above could be helpful).

Reply:
Do you think this one is needed since I've mentioned it in the code: https://github.com/malloc-42/malloc-42.github.io/pull/5/files#diff-d79c25a8390a2547311c733d694e470a3ee96bdad3c195e8a77528e5ae7bbb10R70-R71.

Reply:
Ah missed that. Upto you :) Thanks